src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -868,98 +1128,1238 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
-                            "waiting_for_pick_batches_fail");
+  if (yield_call_combiner) {
+    if (num_batches > 0) {
+      grpc_transport_stream_op_batch_finish_with_failure(
+          exec_ctx, batches[0], GRPC_ERROR_REF(error), calld->call_combiner);
+    } else {
+      GRPC_CALL_COMBINER_STOP(exec_ctx, calld->call_combiner,
+                              "pending_batches_fail");
+    }
   }
   GRPC_ERROR_UNREF(error);
 }
 
 // This is called via the call combiner, so access to calld is synchronized.
-static void run_pending_batch_in_call_combiner(grpc_exec_ctx *exec_ctx,
-                                               void *arg, grpc_error *ignored) {
-  call_data *calld = (call_data *)arg;
-  if (calld->waiting_for_pick_batches_count > 0) {
-    --calld->waiting_for_pick_batches_count;
-    grpc_subchannel_call_process_op(
-        exec_ctx, calld->subchannel_call,
-        calld->waiting_for_pick_batches[calld->waiting_for_pick_batches_count]);
-  }
+static void resume_pending_batch_in_call_combiner(grpc_exec_ctx *exec_ctx,
+                                                  void *arg,
+                                                  grpc_error *ignored) {
+  grpc_transport_stream_op_batch *batch = (grpc_transport_stream_op_batch *)arg;
+  grpc_subchannel_call *subchannel_call =
+      (grpc_subchannel_call *)batch->handler_private.extra_arg;
+  grpc_subchannel_call_process_op(exec_ctx, subchannel_call, batch);
 }
 
 // This is called via the call combiner, so access to calld is synchronized.
-static void waiting_for_pick_batches_resume(grpc_exec_ctx *exec_ctx,
-                                            grpc_call_element *elem) {
+static void pending_batches_resume(grpc_exec_ctx *exec_ctx,
+                                   grpc_call_element *elem) {
   channel_data *chand = (channel_data *)elem->channel_data;
   call_data *calld = (call_data *)elem->call_data;
+  if (calld->enable_retries) {
+    start_retriable_subchannel_batches(exec_ctx, elem, GRPC_ERROR_NONE);
+    return;
+  }
+  // Retries not enabled; send down batches as-is.
   if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
-    gpr_log(GPR_DEBUG, "chand=%p calld=%p: sending %" PRIuPTR
-                       " pending batches to subchannel_call=%p",
-            chand, calld, calld->waiting_for_pick_batches_count,
-            calld->subchannel_call);
-  }
-  for (size_t i = 0; i < calld->waiting_for_pick_batches_count; ++i) {
-    GRPC_CLOSURE_INIT(&calld->handle_pending_batch_in_call_combiner[i],
-                      run_pending_batch_in_call_combiner, calld,
+    size_t num_batches = 0;
+    for (size_t i = 0; i < GPR_ARRAY_SIZE(calld->pending_batches); ++i) {
+      if (calld->pending_batches[i].batch != NULL) ++num_batches;
+    }
+    gpr_log(GPR_DEBUG, "chand=%p calld=%p: starting %" PRIuPTR
+                       " pending batches on subchannel_call=%p",
+            chand, calld, num_batches, calld->subchannel_call);
+  }
+  grpc_transport_stream_op_batch
```

Comment:
This logic looks copy pasted from the num_batch loops in pending_batches_fail. Could consider pulling into helper function, but use your discretion here because of [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))